### PR TITLE
Fix/link pasting

### LINKED ISSUE #3634: When pasting a link using the browser context menu, nothing happens
### What’s the bug you are facing?

When pasting a link using the browser context menu, nothing happens. 

### Which browser was this experienced in? Are any special extensions installed?

Chrome - No

### How can we reproduce the bug on our side?

1) Add some other content to the editor (e.g. blockquote)
2) Copy a link such as https://github.com/
3) Right click to open browser context menu and then click on Paste
4) Nothing happen

https://user-images.githubusercontent.com/6856285/212952802-5aab14c3-252e-4aa9-af72-1e0326a8f471.mov

Note: This works when using Ctrl + V. It also works for the non-link texts.

### Can you provide a CodeSandbox?

https://codesandbox.io/s/ffg1k

### What did you expect to happen?

The link should be pasted successfully when clicking on Paste. The fact that nothing happens is very confusing.

### Anything to add? (optional)

_No response_

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

### LINKED ISSUE #4484: [Bug]: Link pasting unexpected behavior change
### Which packages did you experience the bug in?

core, extension-link

### What Tiptap version are you using?

2.1.11

### What’s the bug you are facing?

Pasting a link on existing text used to add a link mark on the text while keeping the text, it now seems to replace the text with the link.

### What browser are you using?

Chrome

### Code example

https://tiptap.dev/api/marks/link

### What did you expect to happen?

Pasting a link on selected text will just add a link mark on the selected text.

### Anything to add? (optional)

This used to work in 2.0.4.

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

### LINKED ISSUE #4541: [Bug]: Pasting text with more than one link sets all links' href to the first link's href
### Which packages did you experience the bug in?

link

### What Tiptap version are you using?

2.1.12

### What’s the bug you are facing?

I am trying to copy-paste a piece of content that contains more than one link, and the links' text content can be interpreted as valid URLs. Below an example of such content:

--- start copying below ---

Some search engines:
- DuckDuckGo: [https://duckduckgo.com/](https://duckduckgo.com/)
- Bing: [https://www.bing.com/](https://www.bing.com/)

--- stop copying above ---

When I copy-paste this to a TipTap editor, every link but the first one has a wrong `href` attribute - all links have a `href` attribute equal to the `href` attribute of the first link. See a screen recording below.

Basically on paste, TipTap changes this:
```
Some search engines:
- DuckDuckGo: <a href="https://duckduckgo.com/">https://duckduckgo.com/</a>
- Bing: <a href="https://www.bing.com/">https://www.bing.com/</a>
```

To this:
```
Some search engines:
- DuckDuckGo: <a href="https://duckduckgo.com/">https://duckduckgo.com/</a>
- Bing: <a href="https://duckduckgo.com/">https://www.bing.com/</a>
```

This only happens for links whose text content is also an URL - it doesn't happen when links are for example like `[DGG](https://duckduckgo.com/)`

### What browser are you using?

Firefox

### Code example

_No response_

### What did you expect to happen?

I expect all links to always have the same href attribute as on the source page from which I copied them.

### Anything to add? (optional)

https://github.com/ueberdosis/tiptap/assets/7852553/8a3f6431-a39a-4cc1-95e0-c67e6057fcd8

I didn't prepare a CodeSandbox because this behavior can be observed right now on TipTap's official page for the link extension, as seen on the video.

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [X] Yes, I’m a sponsor. 💖

### LINKED ISSUE #4627: [Bug]:
### Which packages did you experience the bug in?

link

### What Tiptap version are you using?

1.32.2

### What’s the bug you are facing?

If I have multiple links in my clipboard and paste them, the output is this:
image

They all get wrapped with html elements in one anchor..

### What browser are you using?

Chrome

### Code example

_No response_

### What did you expect to happen?

I expected each pasted link to be in it's own anchor, the html elements like <br> between the anchor html elements.

### Anything to add? (optional)

_No response_

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

### PR DESCRIPTION
**Pull request by bdbch**
## Please describe your changes

This PR fixes issues with paste handling and autolinking with the current version of the Link extension. I refactored the pasting and autolink plugins, reverted my changes from a few months ago to the one before and updated this to support all cases I wanted to fix back then.

To make my changes happen I had to add the clipboard event to the PasteRule handlers. I think this could also be benefitial for other users in the future.

## How did you accomplish your changes

1. Update the PasteHandler to pass through the clipboard event
2. Use the clipboard event to get pasted link data via DOMParser
3. Insert pasted links including their original URLs into the specific positions of the editor
4. Updated autolink to work with this new approach

## How have you tested your changes

I added necessary tests - but you can also check the feature on the Link preview

## How can we verify your changes

I added necessary tests - but you can also check the feature on the Link preview

## Checklist

- [x] The changes are not breaking the editor
- [x] Added tests where possible
- [x] Followed the guidelines
- [x] Fixed linting issues

## Related issues

closes #4627 
closes #4541 
closes #4484 
closes #3634

### CODE REVIEW

**Comment by duhaime** [comment]
@bdbch looks like there are some failing tests, have you looked into those?

**Comment by bdbch** [comment]
Just figured out that there a few small bugs with link pasting and validation when using the validate option. I now work on this and will probably get this out too tomorrow.

**Comment by C-Hess** [comment]
Looking at this, it looks like it would resolve the need for #4587 as well?

**Comment by bdbch** [comment]
@C-Hess I think it does. I'll check again afterwards

File: packages/core/src/helpers/getMarksBetween.ts [hunk 1]
```diff
@@ -26,6 +26,10 @@ export function getMarksBetween(from: number, to: number, doc: ProseMirrorNode):
       })
   } else {
     doc.nodesBetween(from, to, (node, pos) => {
+      if (!node || node.nodeSize === undefined) {
```
Comment (stevobm) [inline packages/core/src/helpers/getMarksBetween.ts:29]: This change caused a bug: https://github.com/ueberdosis/tiptap/issues/4785

**Review by stevobm — COMMENTED** [review]
Wish to fix the bug caused from here.

**Pull request merged**